src/kontrol/kdist/assert.md

```diff
@@ -49,6 +50,14 @@ These macros expand into the `#assert` production defined earlier, providing an
     rule #assert_not_eq W1 W2 ERR => #assert (W1 =/=Int W2) ERR +Bytes String2Bytes(": " +String Int2String(W1) +String " == " +String Int2String(W2))
     rule #assert_approx_eq_abs W1 W2 W3 ERR => #assert ((maxInt(W1, W2) -Int minInt(W1, W2)) <=Int W3) ERR 
       +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2) +String " (max delta: " +String Int2String(W3) +String ", real delta: " +String Int2String(maxInt(W1, W2) -Int minInt(W1, W2)) +String ")")
+    rule #assert_approx_eq_rel W1 W2 W3 ERR => #assert (W1 ==Int W2) ERR 
+      +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2) 
+      +String " (max delta: " +String Int2String(W3 divInt (10 ^Int 16)) +String "% , real delta: undefined)") // displayed percentage in Kontrol does not include any decimal values
+      requires W2 ==Int 0
+    rule #assert_approx_eq_rel W1 W2 W3 ERR => #assert ((((maxInt(W1, W2) -Int minInt(W1, W2)) *Int (10 ^Int 18)) divInt absInt(W2)) <=Int W3) ERR 
+      +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2) 
+      +String " (max delta: " +String Int2String(W3 divInt (10 ^Int 16)) +String "% , real delta: " +String Int2String((((maxInt(W1, W2) -Int minInt(W1, W2)) *Int (10 ^Int 18)) divInt absInt(W2)) divInt (10 ^Int 16)) +String "%)") // displayed percentage in Kontrol does not include any decimal values
```

Comment:
Do you think there might be a way to compute and cache the delta once, to avoid computing it twice — once when checking if the assert holds, and another — to generate an error string?

Reply:
Oh. This is a good point! Let me think about how to do it. If you have examples that somehow caches values, that would be good for me to learn from it.

Reply:
I think it should be possible to introduce a variable to use within the rule, but I'm not sure how — maybe you can ask in K or Kontrol channel? ty!

Reply:
You could try using the `#as` pattern ([link here](https://github.com/runtimeverification/k/blob/9100689e357a7e5820362869a73684468ef9c91b/docs/user_manual.md?plain=1#L2086))

Reply:
Alternatively, you can rewrite the `#assert_approx_eq_rel W1 W2 W3 ERR` rule into an aux rule that has the delta already computed.
Something like
```k
<k> #assert_approx_eq_rel W1 W2 W3 ERR => #assert_approx_eq_rel_aux W1 W2 W3 divInt (10 ^Int 16) ... </k>
```

Or you could compute the max-delta in the `#cheatcode_call` rule like
```k
    rule [cheatcode.call.assertApproxEqRel]:
         <k> #cheatcode_call SELECTOR ARGS => #assert_approx_eq_rel _ _ (#asWord(#range(ARGS, 64, 32)) divInt (10 ^Int 16) ... </k>
      requires ...

```

Reply:
Thanks for all the suggestions. I will take all these into considerations and update the PR.

Reply:
Hi @palinatolmach and @anvacaru. After experimenting with different methods for the `#assert_approx_eq_rel`, here are my findings:

I ran the following test:
```solidity
    function test_assertApproxEqRel_uint_vA1(uint256 a, uint256 b, uint256 maxPercentDelta) public pure 
    {
        if (b == 0) {
            a = 0;
        }
        else {
            uint256 absDelta = max(a, b) - min(a, b);
            vm.assume(absDelta <= type(uint256).max / 1e18); // prevent overflow of absDelta * 1e18
            uint256 percentDelta = absDelta * 1e18 / b;
            vm.assume(maxPercentDelta >= percentDelta);
        }
        assertApproxEqRel(a, b, maxPercentDelta);
    }
```

with the following 3 versions of 

1. Current PR version:
    ```k
        rule #assert_approx_eq_rel W1 W2 W3 ERR => #assert (W1 ==Int W2) ERR 
          +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2) 
          +String " (max delta: " +String Int2String(W3 divInt (10 ^Int 16)) +String "% , real delta: undefined)") // displayed percentage in Kontrol does not include any decimal values
          requires W2 ==Int 0
        rule #assert_approx_eq_rel W1 W2 W3 ERR => #assert ((((maxInt(W1, W2) -Int minInt(W1, W2)) *Int (10 ^Int 18)) divInt absInt(W2)) <=Int W3) ERR 
          +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2) 
          +String " (max delta: " +String Int2String(W3 divInt (10 ^Int 16)) +String "% , real delta: " +String Int2String((((maxInt(W1, W2) -Int minInt(W1, W2)) *Int (10 ^Int 18)) divInt absInt(W2)) divInt (10 ^Int 16)) +String "%)") // displayed percentage in Kontrol does not include any decimal values
          requires W2 =/=Int 0   
    ```

2. Using `#assert_approx_eq_rel_aux`:
    ```k
        rule #assert_approx_eq_rel W1 W2 W3 ERR => #assert (W1 ==Int W2) ERR 
          +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2) 
          +String " (max delta: " +String Int2String(W3 divInt (10 ^Int 16)) +String "% , real delta: undefined)") // displayed percentage in Kontrol does not include any decimal values
          requires W2 ==Int 0
        rule #assert_approx_eq_rel W1 W2 W3 ERR => #assert_approx_eq_rel_aux W1 W2 (((maxInt(W1, W2) -Int minInt(W1, W2)) *Int (10 ^Int 18)) divInt absInt(W2)) W3 ERR 
          requires W2 =/=Int 0   
        rule #assert_approx_eq_rel_aux W1 W2 W W3 ERR => #assert (W <=Int W3) ERR 
          +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2) 
          +String " (max delta: " +String Int2String(W3 divInt (10 ^Int 16)) +String "% , real delta: " +String Int2String(W divInt (10 ^Int 16)) +String "%)") // displayed percentage in Kontrol does not include any decimal values
    ```

3. Using `#let ...#in`:
    ```k
        rule #assert_approx_eq_rel W1 W2 W3 ERR => 
          #let W3_D = W3 divInt (10 ^Int 16) #in
          #assert (W1 ==Int W2) ERR 
          +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2) 
          +String " (max delta: " +String Int2String(W3_D) +String "% , real delta: undefined)") // displayed percentage in Kontrol does not include any decimal values
          requires W2 ==Int 0
        rule #assert_approx_eq_rel W1 W2 W3 ERR => 
          #let W = ((maxInt(W1, W2) -Int minInt(W1, W2)) *Int (10 ^Int 18)) divInt absInt(W2) #in
          #let W3_D = W3 divInt (10 ^Int 16) #in
          #let W_D = W divInt (10 ^Int 16) #in
          #assert (W <=Int W3) ERR 
          +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2) 
          +String " (max delta: " +String Int2String(W3_D) +String "% , real delta: " +String Int2String(W_D) +String "%)") // displayed percentage in Kontrol does not include any decimal values
          requires W2 =/=Int 0   
    ```

It seems like the current version is the fastest (average of 4m45s) surprisingly, while using `#assert_approx_eq_rel_aux` and `#let ... #in` took around on average 5m15s and 5m30s respectively. I believe the other 2 might be fastest if we can test for the log string output if the assertion fails.

Should I stick to the current version or change to any of the other 2 versions (as the other 2 versions seem more presentable)?

Reply:
@jinxinglim 
One more comment, I would suggest formatting the rules like this to increase readability. In addition, the comments can go outside the ```k block.
```
    rule #assert_approx_eq_rel W1 W2 W3 ERR => #assert (W1 ==Int W2) ERR 
                                        +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2)
                                                                 +String " (max delta: " +String Int2String(W3 divInt (10 ^Int 16))
                                                                 +String "% , real delta: undefined)")
      requires W2 ==Int 0

    rule #assert_approx_eq_rel W1 W2 W3 ERR => #assert ((((maxInt(W1, W2) -Int minInt(W1, W2)) *Int (10 ^Int 18)) divInt absInt(W2)) <=Int W3) ERR
                                        +Bytes String2Bytes(": " +String Int2String(W1) +String " !~= " +String Int2String(W2)
                                                                 +String " (max delta: " +String Int2String(W3 divInt (10 ^Int 16)) +String "% , real delta: " 
                                                                 +String Int2String((((maxInt(W1, W2) -Int minInt(W1, W2)) *Int (10 ^Int 18)) divInt absInt(W2)) divInt (10 ^Int 16)) +String "%)")
      requires W2 =/=Int 0
```

Reply:
Okay. Formatted in this commit https://github.com/runtimeverification/kontrol/pull/593/commits/34ebaa28f2c313f9315b6eb56eedaa4d198f21e5. Thank you!